pkg/activator/throttler.go

```diff
@@ -0,0 +1,191 @@
+type Throttler struct {
+	breakers      map[RevisionID]*queue.Breaker
+	breakerParams queue.BreakerParams
+	logger        *zap.SugaredLogger
+	getEndpoints  func(RevisionID) (int32, error)
+	getRevision   func(RevisionID) (*v1alpha1.Revision, bool, error)
+	mux           sync.Mutex
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity updates the max concurrency of the Breaker corresponding to a revision.
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, err := t.updateCapacity(rev, size)
+	return err
+}
+
+// This method updates Breaker's concurrency and requires external synchronization, e.g. use the mux.
+// It create a new breaker and saves it into our bookkeeping if doesn't exist.
+// This is important for not loosing the update signals
+// that came before the requests reached the Activator's Handler.
+func (t *Throttler) updateCapacity(rev RevisionID, size int32) (breaker *queue.Breaker, err error) {
+	revision, exists, err := t.getRevision(rev)
+	if err != nil {
+		return breaker, err
+	}
+	if !exists {
+		return breaker, fmt.Errorf("no revision was found for: %s", rev)
+	}
+	cc := int32(revision.Spec.ContainerConcurrency)
+	// The concurrency is unlimited, thus hand out as many tokens as we can in this breaker.
+	if cc == 0 {
+		cc = t.breakerParams.MaxConcurrency
+	}
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		breaker = queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = breaker
+	}
+	delta := size*cc - breaker.Capacity()
+	// Do not update throttler's concurrency if the same number of endpoints is received
+	// or the number is negative.
+	if delta <= 0 {
+		return breaker, nil
+	}
+	if err := breaker.UpdateConcurrency(delta); err != nil {
+		return breaker, err
+	}
+	return breaker, nil
+}
+
+// Try potentially registers a new breaker in our bookkeeping
+// and executes the `function` on the Breaker.
+// It returns an error if either breaker doesn't have enough capacity,
+// or breaker's registration didn't succeed, e.g. getting endpoints or update capacity failed.
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	breaker, err := t.getOrCreateBreaker(rev)
+	if err != nil {
+		return err
+	}
+	if ok := breaker.Maybe(function); !ok {
+		return errors.New(OverloadMessage)
+	}
+	return nil
+}
+
+// Get existing breaker or create a new one.
+// In the latter case fetch the endpoints and update the capacity of the newly created breaker.
+// This avoids a potential deadlock in case if we missed the updates from the Endpoints informer.
+// This could happen because of a restart of the Activator or when a new one is added as part of scale out.
+func (t *Throttler) getOrCreateBreaker(rev RevisionID) (breaker *queue.Breaker, err error) {
```

Comment:
I'd propose a decomposition of `getOrCreateBreaker` and take the "update the state now" parts of it out into a separate function. That makes it crystal clear that `Try` might actually fetch state from the API etc. etc.

I sketched such a decomposition here: https://github.com/knative/serving/commit/8307fbb5e578a65106af19b0a33d830c7e1abd86

I feel that puts "getOrCreate" calls in more clear spaces and in general has less hidden defaulting logic. WDYT?